Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataServiceObserver.cpp

```diff
@@ -15,26 +16,71 @@ using namespace Mantid::API;
 using namespace Mantid::PythonInterface;
 using namespace boost::python;
 
+namespace {
+void observeAll(AnalysisDataServiceObserver &self, bool turnOn) {
```

Comment:
I was wondering whether we might be able to reduce the code duplication here using a more generic function and `std::bind`, for example:

```cpp
#include <functional>
#include <iostream>

// Example class as a substitute for AnalysisDataServiceObserver
struct Observer {

  void hello(bool on) {
     std::cerr << "Hello\n";
  }

  void bye(bool on) {
     std::cerr << "Bye\n";
  }

};

// Function pointer to a method with this signature
using ObserverMethod = void(Observer::*)(bool);

// Single function that calls the given method
void callReleasingGIL(Observer &self, bool on, ObserverMethod method) {
  // RELEASE GIL
  (self.*method)(on);
}

int main() {
  // use std::bind to create functions that are bound to the correct member function
  using namespace std::placeholders;
  std::bind(callReleasingGIL, _1, _2, &Observer::hello);
  std::bind(callReleasingGIL, _1, _2, &Observer::bye);

  // I am imagining something like the following below:
  // .def("observeAll", std::bind(callReleasingGIL, _1, _2, &AnalysisDataServiceObserver::observeAll);

}
```

An example on  [compiler explorer](https://gcc.godbolt.org/z/o3PzKGq5b)

Reply:
Ah nice thank you! I'll have a go